### ISSUE TITLE
High-value plugin ideas

### ISSUE THREAD

**Issue by jni** [napari/napari#939]
This is a meta-issue to collect ideas for plugins that we want to see sooner rather than later, once we have some plugin infrastructure in place. (Thanks @tlambert03! :grinning:) I'm writing this in preparation for RGSoC, as plugins would make a great development project for students.

- [x] Ilastik-like interactive segmentation
- [x] Segmentation editor: merge labels, split labels with e.g. watershed, label painting
- [x] Manual registration picking correspondence points
- [ ] Manual registration with parameter tuning (e.g. translation and rotation along all axes, plus optional view of a scoring metric)
- [ ] Deep learning viewer a-la tensorboard
- [ ] Deep learning model views (e.g. visualisation of the weights of all the layers of a model)
- [x] Regionprops viewer (polished version of @VolkerH's viewer)
- [x] Matplotlib integration (putting interactive plots in dockable widgets)
- [ ] UI for interactive watershed seed selection as demonstrated with code in the blog post
- [ ] Skeleton viewer + analyzer (integrated with skan)
- [ ] 3D point picking and line tracing
- [ ] viewing 3D+t tracks, + editing?
- [ ] Register translation using skimage.

**Update** 23/10/2020

- [ ] high content screen plate viewer (xref ome/ome-zarr-py#57, ome/omero-ms-zarr#74)

**Comment by VolkerH** [comment]
I'm rather keen to work on the regionprops viewer once a plugin infrastructure is in place. I did not follow all the issues/discussions on this during my absence in January, will read up on the `pluggy` discussions. Possibly @jni can bring me up to speed with this.

**Comment by tlambert03** [comment]
this is awesome and really fun about @jni!  For those reading up on this and thinking about how to implement some of these items, i think the key things to identify are:

1. where exactly in the napari codebase a [hook function](https://pluggy.readthedocs.io/en/latest/#how-does-it-work) needs to be added.
2. what is the signature (both arguments and types) of that function that will be passed to the plugin (i.e. what [hookspec](https://pluggy.readthedocs.io/en/latest/#specifications) would we be adding)
3. what considerations will there be when actually [calling the hooks](https://pluggy.readthedocs.io/en/latest/#calling-hooks), collecting the results, and dealing with exceptions.

**Comment by sofroniewn** [comment]
On PyPi I have grabbed the following names

- napari-classification
- napari-denoising
- napari-detection
- napari-io
- napari-registration
- napari-segmentation
- napari-tracking

My idea for those plugins, would be to try and turn them into their own little plugin ecosystems - i.e. `napari-segmentation` would provide some convenience functionality that all segmentation plugins might want to use and provide hooks for specific segmentation algorithms that are more specific than the hooks provided by napari. I'm not wedded to this strategy, but thought it might be an approach of interest  further down the line. I see this approach as complementary to people who want to make a full on specific segmentation plugin themselves, and is something that might come much later for us, after some initial plugins get made from the list @jni mentions above.

**Comment by sofroniewn** [comment]
- [ ] add a reader plugin for the "surface" layer that uses https://github.com/nschloe/meshio - see https://github.com/vispy/vispy/issues/1817
- [ ] [AICSImageIO](https://github.com/AllenCellModeling/aicsimageio) - a Python library for reading and writing image data with special handlers for bio-formats. See more discussion here https://forum.image.sc/t/aicsimageio-3-1-0-release/34343

**Comment by adamltyson** [comment]
Once there is a plugin infrastructure is in place, there are a few of us who would like to contribute to a registration plugin for napari (cc @FedeClaudi). We're particularly interested in registration of 2D sections to a 3D volume, which would likely involve a lot of manual tweaking (and napari seems perfect for).

**Comment by MMathisLab** [comment]
One vote for DeepLabCut for multi-keypoints tracking of animals, cells, etc: 
www.deeplabcut.org

**Comment by tlambert03** [comment]
We still need to write up developer-focused plugin documentation, but I'd just like to add a little background info here for any developers thinking ahead:

For now at least (since https://github.com/napari/napari/pull/908), what it means to "plug in" to napari is to write a function (called a "hook implementation") that will get called at some specific place (a "hook") inside of the napari codebase.  This means that for each new "type" of plugin that we want to offer support for, we need to do two things:

1. Determine the _exact_ function signature that a plugin developer will need to implement: i.e. exactly what arguments will the hook function accept, what are the argument types, and what type of object will the hook return.  These signatures are called "hook specifications" and as they are added, you will find them in [`napari.plugins.hook_specifications`](https://github.com/napari/napari/blob/master/napari/plugins/hook_specifications.py).

    > for an example hook specification see [`napari_get_reader`](https://github.com/napari/napari/blob/bd538488c9a6f9d6721cec8268610a1ea5b0c7d0/napari/plugins/hook_specifications.py#L33-L56)

2. Determine the _exact_ line(s) in the napari codebase where that hook function will be called, and what exactly napari is supposed to do with the return value.

    > for example, with `napari_get_reader`, see [this (work-in-progress) hook](https://github.com/napari/napari/blob/a7c7da78fa1f68bc6de9fe0e669d17fbb5762d11/napari/plugins/utils.py#L100-L153)

If you are a developer who is particularly excited to extend napari functionality, we welcome you to familiarize yourself with the [hook/hook_specification/hook_implementation pattern](https://pluggy.readthedocs.io/en/latest/), and think about where in the napari codebase the hook might go.  As you come up with ideas, please open [new issues](https://github.com/napari/napari/issues/new/choose) or PRs to propose additional places where napari might offer hooks!

**Comment by neuromusic** [comment]
@jni can you expand on why these plugins should be "sooner than later"? because they are foundational and every bioimage analyst will want to use them? because they are a "flagship" that demonstrate how napari differentiates from competing image visualization+analysis tools? I'm curious how you are thinking about this and you would approach prioritization.

**Comment by VolkerH** [comment]
Regarding plugins in general, if they add new functionality they will typically bring new dependencies. I believe the napari developers are trying to keep unnecessary dependencies at a minimum (at least that is my impression from interactions with @jni). However, I assume that the plugins have to live in the same python environment as the viewer. I expect some of the highly desirable plugins to come with dependencies such as tensorflow (or tensorflow-gpu), pandas etc. 

As the plugins probably can't live in separate python environments, are there any ideas on how to address this? I fear the dependency hell trying to create an environment in which napari and multiple plugins from different groups have to coexist. Not trivial to manage for intermediate python users and almost impossible for end-users. 

To some degree, this is the problem that Cellprofiler faces with their plugins as well. There are some high-value plugins (segmentation using unet) but they don't work with the standard binary distribution of Cellprofiler. One has to create an environment in which both Cellprofiler and the plugin run. 

Feel free to separate this thought from the thread where plugin ideas are collected - maybe it should have gone elsewehere, I just don't want to create too many new issues.

**Comment by tlambert03** [comment]
Super good point 👆.  Dependency hell will a very real issue with multiple plugins with heavy dependencies. Thanks @VolkerH. Possible that conda could help slightly here, but some users will likely still run into unsolveable dependencies between plugins (and the plugins would also need to have conda installs of course)

**Comment by VolkerH** [comment]
Just thought of imjoy. The plugin architecture there is based on plugins running in their own process and exchanging data with the main program via rpcs. That has some limitations when it comes to tight integration with the viewer (I'm thinking e.g.  of. GUI elements ot direct access to GPU functionality via vispy) but avoids the depedency hell.

**Comment by tlambert03** [comment]
just thinking out loud... if we did eventually want to go that direction (RPCs and messages) we could, it would just require moving away from pluggy and rolling our own plugin registration system (since pluggy is pretty much built on the assumption that you can import the plugin module at registration time).  That's might not be _that_ big of a job though, and we can definitely take what we learn from the pluggy hooks/spec/implementation pattern when doing so... (if it comes to that).

**Comment by sofroniewn** [comment]
> Feel free to separate this thought from the thread where plugin ideas are collected - maybe it should have gone elsewehere, I just don't want to create too many new issues.

I just pulled of the discussion from @VolkerH and @tlambert03 on plugin dependency management into its own issue #1001 and it would be great if we could keep discussion on that there as I do see it as separate from ideas on new plugins.

@VolkerH really never worry about creating too many issues, it's actually much easier for me to keep track of more issues that are smaller / more granular and tightly focused as I can then close them when the discussion is done. In the event that you do create a new issue for a something that already had an issue it's also easier to close the duplicate issue and put that comment into the existing issue rather then pull out a discussion like the one above into something new. But really no worries either way, we're just glad you're surfacing such important points for discussion!!

**Comment by neuromusic** [comment]
Do I understand correctly that "plugins" for napari is first and foremost an architecture decision? And that when #376 or some variant is implemented, some fundamental plugins (e.g. i/o) will likely be bundled with the application?

**Comment by jni** [comment]
@neuromusic bundling is its own discussion also. =)

**Comment by jni** [comment]
@neuromusic re "high value" and "sooner", these are applications/workflows that I and others have already received requests for in the context of napari.

**Comment by GenevieveBuckley** [comment]
@VolkerH suggested via email that another good idea would be a plugin similar to AnnotatorJ, for semi-automatic image annotations.

*AnnotatorJ: an ImageJ plugin to ease hand-annotation of cellular compartments*
https://www.biorxiv.org/content/10.1101/2020.02.27.968362v1.full.pdf

**Comment by neuromusic** [comment]
Someone at NEUBIAS asked for a Napari implementation of Stephan Saalfeld's Interactive Moving Least Squares ImageJ plugin to facilitate manual registration of cross-modal imaging data.

https://imagej.net/Interactive_Moving_Least_Squares

**Comment by neuromusic** [comment]
@carsen-stringer's cellpose

https://twitter.com/computingnature/status/1224477812763119617

**Comment by haesleinhuepf** [comment]
How about adding a pythonized CLIJ as plugin for generic image processing? We could call it [clEsperanto](http://clEsperanto.net) 😉

**Comment by sumanthratna** [comment]
Along with segmentation, I think it'd be great if there was a plugin for binary thresholding. That way, if I have a grayscale image, I don't need to do `viewer.add_image(skimage.whatever.import(grayscale_img))` for each algorithm I want to try; instead, I can just click a few buttons and see the outputs of all the algorithms.
